<a href="https://colab.research.google.com/github/narayana1111008989/EDA-hotel_booking_analysis/blob/main/Copy_of_Hotel_Booking_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

## <b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

## <b> Explore and analyze the data to discover important factors that govern the bookings. </b>

# Exploratory Data Analysis on Hotel Bookings

For this analysis a dataset containing the booking information of two hotels is used. Let's import the required libraries and data into a pandas dataframe called hotel_booking_df.

In [ ]:
# Mounting google drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
# Importing csv file from google drive and storing it in a dataframe.
hotel_booking_df = pd.read_csv("/content/drive/MyDrive/Copy of Hotel booking analysis.csv")

In [ ]:
# Let's look at an overview of columns.
hotel_booking_df.head()

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
hotel_booking_df.shape

(119390, 32)

In [ ]:
hotel_booking_df.tail()

In [ ]:
# Checking the number of rows and columns in the dataset.
print(f'This dataset is having {hotel_booking_df.shape[0]} rows and {hotel_booking_df.shape[1]} columns.')

This dataset is having 119390 rows and 32 columns.


Let's look at what is contained in the data and what do they mean.

# Explanation of columns contained in the dataset:


1.   hotel : This dataset contains the data of two hotels. One is a resort hotel and the other is a city hotel. 

2.   is_cancelled : This column contains two values. '0' for did not cancel and '1' for cancelled.

3.   lead_time : Number of days between the customer booking the room and actually arriving.

4.   arrival_date_year : Year of the date the customer arrived.

5.   arrival_date_month : Month of the date the customer arrived.

6.   arrival_date_week_number : Week number in the year of the date the customer arrived.

7.   arrival_date_day_of_month : Day of the date the customer arrived.

8.   stays_in_weekend_nights : Number of days the customer stayed on the weekends.

9.   stays_in_week_nights : Number of days the customer stayed on week nights.

10.  adults : Number of adults per booking.

11.  children : Number of children per booking.

12.  babies : Number of babies per booking.

13.  meal : Type of meal ordered. The categories are
         undefined/SC - no meal package
         BB – Bed & Breakfast
         HB – Half board (breakfast and one other meal – usually dinner)
         FB – Full board (breakfast, lunch and dinner)

14.  country : Nationality of the customer.

15.  market_segment : Specifies from where the room was booked,
         Direct - By the customer without any third party.
         Corporate - By the customers company.
         Online TA - By an online travel agent.
         Offline TA/TO - By a traditional travel agent or 
                         a tour operator.
         Complementary - Booked as a gift.
         Groups - Customers who are there for events such 
         as conferences and not just for stays.
         Aviation - Pilots and crew.

16.  distribution_channel : Specifies how the room was booked,
         Corporate - by the company.
         Direct - by an individual.
         TA/TO - by a Travel Agency or Tour Operator.
         GDS - Global Distribution System(corporates use these 
               technology companies for booking for their staff)
                  
17.  is_repeated_guest : Value 1 for repeated guest and 0 for new guests.

18.  previous_cancellations : Number of previous bookings that were cancelled  by the customer prior to the current booking.

19.  previous_bookings_not_canceled : Number of previous bookings not cancelled by the customer prior to the current booking.

20.  reserved_room_type : Code of room type reserved.

21.  assigned_room_type : Code for the type of room assigned to the booking.

22.  booking_changes : 	Number of changes/amendments made to the booking from the moment the booking was made until the moment of check-in or cancellation.

23.  deposit_type : Whether the customer gave an advance deposit,
         No Deposit – no deposit was made.
         Non Refund – a deposit was made in the value of the 
         total stay cost.
         Refundable – a deposit was made with a value under 
                      the total cost of stay.

24.  agent : ID of the travel agency that made the booking.

25.  company : ID of the company/entity that made the booking or responsible for paying the booking. 

26.  days_in_waiting_list : Number of days the booking was in the waiting list before it was confirmed to the customer.

27.  customer_type : Type of customer, assuming one of four categories,
         Contract - arrangement between the hotel and a company
         Group – the customers did not come to stay per se but 
                 for an event, like a conference.
         Transient – the customer is a walk in or spontaneous.
         Transient-party – a group of transient customers.

28.  adr : Average Daily Rate, calculated by dividing the sum of all lodging transactions by the total number of staying nights.

29.  required_car_parking_spaces : Number of car parking spaces required by the customer.

30.  total_of_special_requests : Number of special requests made by the customer (e.g. twin bed or high floor).

31.  reservation_status : Reservation last status,
         Canceled – booking was canceled by the customer;
         Check-Out – customer has checked in but 
                     already departed;
         No-Show – customer did not check-in and did 
                   inform the hotel of the reason why

32.  reservation_status_date : Date at which the last status was set.

Purging the dataset of null values.

In [ ]:
# To get an idea of the amount of null values in every column
hotel_booking_df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

Four columns have null values. Columns agent and company seem to have a comparitively higher number of null values. To get an idea of how high this is we can divide by the total number of rows which is 119390.

In [ ]:
# To find the percentage of null values out of the total number of rows.
hotel_booking_df[['children', 'country', 'agent', 'company']].isnull().sum() / hotel_booking_df.shape[0]

children    0.000034
country     0.004087
agent       0.136862
company     0.943069
dtype: float64

In the columns children and country only 0.003 % and 0.41 % have null values. This is negligible. Therefore, the corresponding rows can be ignored. However, for the columns agent and company the percentage of null rows is relatively high. Trustworthy insights cannot be obtained in such conditions hence we delete these columns.

In [ ]:
# Deleting two columns and deleting rows which contains null values.
hotel_booking_df = hotel_booking_df.drop(['agent', 'company'], axis = 1)
hotel_booking_df = hotel_booking_df.dropna(axis = 0)

There are also some instances of 'undefined'.

In [ ]:
# Checking which columns have 'Undefined' in them.
(hotel_booking_df == 'Undefined').any()

hotel                             False
is_canceled                       False
lead_time                         False
arrival_date_year                 False
arrival_date_month                False
arrival_date_week_number          False
arrival_date_day_of_month         False
stays_in_weekend_nights           False
stays_in_week_nights              False
adults                            False
children                          False
babies                            False
meal                               True
country                           False
market_segment                    False
distribution_channel               True
is_repeated_guest                 False
previous_cancellations            False
previous_bookings_not_canceled    False
reserved_room_type                False
assigned_room_type                False
booking_changes                   False
deposit_type                      False
days_in_waiting_list              False
customer_type                     False


We can see that there are two columns which have 'undefined' values, meals and distribution_channel. 'Undefined' in meals suggests that the customer did not opt for any food from the hotel. The code in meals column 'SC' which stands for self catered also means the same. Therefore, can change 'undefined' to 'SC'.

In [ ]:
hotel_booking_df['meal'].value_counts()

BB           91863
HB           14434
SC           10638
Undefined     1165
FB             798
Name: meal, dtype: int64

In [ ]:
hotel_booking_df['meal'].replace({'Undefined' : 'SC'}, inplace = True)

In [ ]:
hotel_booking_df['distribution_channel'].value_counts()

TA/TO        97730
Direct       14483
Corporate     6491
GDS            193
Undefined        1
Name: distribution_channel, dtype: int64

In [ ]:
hotel_booking_df = hotel_booking_df[hotel_booking_df['distribution_channel'].str.contains('Undefined')==False]

In [ ]:
hotel_booking_df.head(10)

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   
5  Resort Hotel            0         14               2015               July   
6  Resort Hotel            0          0               2015               July   
7  Resort Hotel            0          9               2015               July   
8  Resort Hotel            1         85               2015               July   
9  Resort Hotel            1         75               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   
5                        27                          1   
6                        27                          1   
7                        27                          1   
8                        27                          1   
9                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
0                        0                     0       2  ...   
1                        0                     0       2  ...   
2                        0                     1       1  ...   
3                        0                     1       1  ...   
4                        0                     2       2  ...   
5                        0                     2       2  ...   
6                        0                     2       2  ...   
7                        0                     2       2  ...   
8                        0                     3       2  ...   
9                        0                     3       2  ...   

   assigned_room_type  booking_changes deposit_type days_in_waiting_list  \
0                   C                3   No Deposit                    0   
1                   C                4   No Deposit                    0   
2                   C                0   No Deposit                    0   
3                   A                0   No Deposit                    0   
4                   A                0   No Deposit                    0   
5                   A                0   No Deposit                    0   
6                   C                0   No Deposit                    0   
7                   C                0   No Deposit                    0   
8                   A                0   No Deposit                    0   
9                   D                0   No Deposit                    0   

  customer_type    adr  required_car_parking_spaces  \
0     Transient    0.0                            0   
1     Transient    0.0                            0   
2     Transient   75.0                            0   
3     Transient   75.0                            0   
4     Transient   98.0                            0   
5     Transient   98.0                            0   
6     Transient  107.0                            0   
7     Transient  103.0                            0   
8     Transient   82.0                            0   
9     Transient  105.5                            0   

   total_of_special_requests  reservation_status reservation_status_date  
0                          0           Check-Out              2015-07-01  
1                          0           Check-Out              2015-07-01  
2                          0           Check-Out              2015-07-02  
3                   

In [ ]:
df=pd.DataFrame(hotel_booking_df.groupby(['hotel', 'is_canceled']).size())
df

0
hotel        is_canceled       
City Hotel   0            46226
             1            33076
Resort Hotel 0            28518
             1            11077

In [ ]:
plt.bar(city hotel,Resort hotel)
plt.ylabel

In [ ]:

df=hotel_booking_df.country
df

0         PRT
1         PRT
2         GBR
3         GBR
4         GBR
         ... 
119385    BEL
119386    FRA
119387    DEU
119388    GBR
119389    DEU
Name: country, Length: 118897, dtype: object

In [ ]:
df.value_counts(normalize=True)

PRT    0.408631
GBR    0.102013
FRA    0.087597
ESP    0.072062
DEU    0.061288
         ...   
DJI    0.000008
BWA    0.000008
HND    0.000008
VGB    0.000008
NAM    0.000008
Name: country, Length: 177, dtype: float64

In [ ]:

MS=hotel_booking_df.market_segment
MS

0                Direct
1                Direct
2                Direct
3             Corporate
4             Online TA
              ...      
119385    Offline TA/TO
119386        Online TA
119387        Online TA
119388        Online TA
119389        Online TA
Name: market_segment, Length: 118897, dtype: object

In [ ]:
MS.value_counts(normalize=True)

Online TA        0.474377
Offline TA/TO    0.203201
Groups           0.166581
Direct           0.104687
Corporate        0.042987
Complementary    0.006173
Aviation         0.001993
Name: market_segment, dtype: float64

In [ ]:
 pd.DataFrame(hotel_booking_df.market_segment)